### PR TITLE
Complete Meals DB App Functionality with Features required

### PR DESCRIPTION
**Pull request by Abdulwhbhussain**
**JavaScript capstone project - Your API-based webapp**
Learning objectives
Use JavaScript to make websites dynamic and build basic single-page apps.
Use ES6 syntax.
Use ES6 modules.
Use callbacks and promises.
Use webpack.
Apply JavaScript best practices and language style guides in code.
Use the AAA pattern for unit tests.
Write unit tests for a JavaScript app.
Follow Gitflow.
Solve simple git conflicts.
Send and receive data from an API.
Use API documentation.
Understand and use JSON.
Make JavaScript code asynchronous.
Perform a code review for a team member.
Estimated time: 20.5h
Description
The JavaScript capstone project ([remember what they are?](https://github.com/microverseinc/curriculum-html-css/blob/main/articles/capstone_intro.md)) is about building your web application based on an external API. You will select an API that provides data about a topic that you like and then build the web app around it. The web app will have 2 or 3 user interfaces (depending on the size of your team):

A home page showing a list of items that you can "like."
A popup window with more data about an item that you can use to comment on it or reserve it for some time.

General requirements
Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
Make sure that you used the correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
HTML/CSS & JavaScript requirements
Follow our list of [best practices for HTML & CSS](https://github.com/microverseinc/curriculum-html-css/blob/main/articles/html_css_best_practices.md).
Follow our list of [best practices for JavaScript](https://github.com/microverseinc/curriculum-html-css/blob/main/articles/javascript_best_practices.md).
Project requirements
IMPORTANT NOTE: in this project, you will also practice giving code reviews to your teammates. Remember to read the [Code Review](https://github.com/microverseinc/curriculum-javascript/blob/main/group-capstone/js_capstone.md#code-review) carefully!

APIs

First, you need to find an API so you can base the development of the web app around it. The API should allow you to:

Get a list of items with a unique item id (or generate the unique id).
For a given item, get detailed information about it.
We recommend that you choose an API that doesn't need authentication. if you select an API that requires authentication, you should implement it on your own. Also, if you select an API that provides image resources your web app will be more visual.

Some example APIs are:

[Pokeapi](https://pokeapi.co/): data about the Pokémon world.
[TVmaze API](https://www.tvmaze.com/api): data about TV series and movies.
[Meals DB](https://www.themealdb.com/api.php): data about meals.
You can find more APIs in [this GitHub repo](https://github.com/public-apis/public-apis) or [ProgrammableWeb](https://www.programmableweb.com/category/all/apis). Some of the APIs require a token, some require authentication, and some others are just open.

You will use our [Involvement API](https://www.notion.so/microverse/Involvement-API-869e60b5ad104603aa6db59e08150270) to record the different user interactions (likes, comments, and reservations).
Interfaces

You should build these interfaces:
The home page.
The comments popup.
The reservations popup (only for groups of 3 students).
You should follow the layout of the wireframes provided. You should personalize the rest of the design including colors, typographies, spacings, etc.
Home page
When the page loads, the web app retrieves data from:
The selected API shows the list of items on the screen.
The Involvement API to show the item likes.
Remember that your page should make only 2 requests:
One to the base API.
And one to the Involvement API.
When the user clicks on the Like button of an item, the interaction is recorded in the Involvement API and the screen is updated.
When the user clicks on the "Comments" button, the Comments popup appears.
When the user clicks on the "Reservations" button, the Reservations popup appears (only for groups of 3 students).
Home page header and navigation are similar to the given mockup.
Home page footer is similar to the given mockup.
Comments popup
When the popup loads, the web app retrieves data from:
The selected API shows details about the selected item.
The Involvement API to show the item comments.
When the user clicks on the "Comment" button, the data is recorded in the Involvement API, and the screen is updated.
Reservations popup (only for groups of 3 students)
When the popup loads, the web app retrieves data from:
The selected API shows details about the selected item.
The Involvement API to show the item reservations.
When the user clicks on the "Reserve" button, the data is recorded in the Involvement API, and the screen is updated.
Counters We have counters in all the interfaces that show:

The number of items (home).
The number of comments (comments popup).
The number of reservations (reservations popup) - only for groups of 3 students.
Even if the API gives you these numbers, you will create a specific function to calculate these numbers on each page. These count functions need to be covered with unit tests using Jest.

Technical set up

Set up the repository on GitHub and use Gitflow.
Set up webpack.
Set up a JavaScript testing library (Jest).
Project documentation
Once you have finished the development of the project, you should record a video presenting the features of the project you built. It is a video with a maximum length of 5 minutes. The content of the video should include:

A description of the project.
A demo of the project features.
You should also highlight some interesting piece of code or something you built that you are very proud of.
You all should appear in the video and talk about the project.
For recording the video you can use the Zoom recording features while in a call with your peers.

Workload distribution
In order to tackle this challenge, we created [a template of the Kanban board with GitHub project](https://github.com/microverseinc/curriculum-javascript/projects/1) that translates the requirements into a set of tasks that you will be able to use to organize your work. You will create your own copy of that board in the separate activity.

You will be working in this way:

The common tasks (set up repo, find the API, shared work on the home page) will be divided among all of you or completed as a team (pair programming).
The tasks related to a specific page will be developed individually, with every student owning the development of one page.
All tasks should be based on the cards from your Kanban board.
Below you can see a suggestion of what you can do every day (just a suggestion, not mandatory).

Day 0
Choose the topic for your website and the API.
Choose images, text, icons, fonts.
Set up the repository and tools.
Day 1
Create the basic shared code on the home page.
Start individual interfaces.
Give code reviews to your teammates.
Day 2
Work on individual interfaces.
Give code reviews to your teammates.
Day 3
Complete the work on individual interfaces.
Record a video for your project.
Create a good README and PR description.
Ask for the external review - using your Student Dashboard.
Work and submission mode
You should implement the above requirements only in one repository in your group.
You should ask for a review and submit this activity on behalf of your group.
Code review
You will give and receive code reviewes from your teammates. Each task (i.e. each card) should have a separate pull request that is reviewed by one of your teammates. Be nice to your teammates and add link to the PR in specific card's comments. Once the entire project is ready, one of your team members will request a code review on behalf of your group. For both proccesses follow [these steps](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/code-review/articles/code_review_flow_group_projects.md).

### CODE REVIEW

File: src/index.js
Comment (brytebee) [inline src/index.js:1]: You're required to define a function that can be tested with unit tests for all counts on the page.

image

I have commented on their occurrences in this file.

This is to ensure that the unit test for the function is easier. Kindly implement the function and add the test.

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    <!-- Food Name -->
+    <div class="popup-name-container"><h2>${foodObject.name}</h2></div>
+    <!-- Food Details -->
+    <div class="popup-details-container">
+    <div class="popup-details-upper"><span>Count: ${foodObject.count}</span><span>ID: ${foodObject.id}</span></div>
+    <div class="popup-details-lower"><span>Likes: ${foodObject.likes}</span><span>Delicious: Yes</span></div>
+    </div>
+    <!-- Comments and comment Details -->
+    <div class="popup-comments-container">
+    <h2>Comments (${foodObject.comments.length})</h2>
```
Comment (brytebee) [inline src/index.js:89]: Here

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            const commentElement = document.createElement('div');
+            commentElement.classList.add('individual-comment');
+            commentElement.innerText = `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()} ${name}: ${comment}`;
+            document.querySelector('.popup-comments').appendChild(commentElement);
+            foodObject.comments.push({
+                comment: `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()} ${name}: ${comment}`,
+                creation_date: `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`,
+                username: name,
+            });
+            document.querySelector('.popup-comments-container h2').innerText = `Comments (${foodObject.comments.length})`;
```
Comment (brytebee) [inline src/index.js:124]: Here

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+document.addEventListener('DOMContentLoaded', async () => {
+  appRender();
+
+  const foodByChina = await foodByChinaList();
+
+  console.log(foodByChina);
+
+  // Show items count on homepage Header
+    const foodCount = document.querySelector('nav > ul > li > a');
+    foodCount.innerHTML = `Chinese(${foodByChina.length})`;
```
Comment (brytebee) [inline src/index.js:140]: Here

**Review by brytebee — CHANGES_REQUESTED** [review]
Hi, @Abdulwhbhussain & @Enoisong 👋, 

##

### Highlights💡
- Good commit message.✔️ 
- Good PR Summary.✔️
- Great UI design.✔️
- App functions as it should.✔️

##

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

Please, remember to tag me in your question so I can receive the notification. 
Please, remember to tag me @brytebee in your question so I can receive the notification.
### _**Use @brytebee on slack for faster responses.**_

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: src/index.js (hunk 3 above)
Comment (Abdulwhbhussain) [reply src/index.js:140]: already tests are there and successfully passing!

**Comment by arslanbisharat** [comment]
Hi @Abdulwhbhussain,
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- The Pull Request has a proper title and description. ✔️ 
- You have added a descriptive Readme file. Good Job. ✔️ 
- There is no linter error is present. ✔️ 
- Project has met all the technical aspects you needed for this project. Keep rocking! ✔️ 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**